### ISSUE TITLE
Lets cross out everything we think is done

### ISSUE THREAD

**Issue by StevenMHernandez** [StevenmHernandez/cmsc312-operating-system-simulator#12]
1. Summary
<strike>This project requires to design and implement a simulation of an operating system. Students are expected to prepare a computer program that will serve as a simulator of loading and execution of programs on a theoretical operating system. These programs will originate from job files created by the programmers. Jobs or simulated programs may consist of a dummy software like word processor, web browser, virus scan software, media player and photo editing software. These programs need not to be coded, only simulated. These job files will have simulated instruction cycles and I/O that is proportional to the type and size of program that it is.
Classes are to be created in order to mimic the role of different hardware components and software embedded in the operating system. These classes will have methods in them to perform operating system duties and will rely or call on other classes or methods, thus creating a fully interconnected system.</strike>

2. Project Description

2.1 Requirements
<strike>The project requires that the hardware consist of one central processing unit and 256 kB of memory. Furthermore, the simulator shall have I/O operations taking between 25 and 50 cycles. The simulator runs in steps defined as “looped” cycles.</strike>

2.2 Features
<strike>A command line embedded within GUI is required, which prompts the user for input. The user will initiate the start of a program with a command line in the simulator. User will specify the number of cycles the program can run for testing & observational purposes. Commands the user can issue are PROC, MEM, LOAD, EXE, RESET, and EXIT.</strike>
- <strike>PROC - shows all unfinished processes in the system and their information. The process information should include: current process state, amount of CPU time needed to complete, amount of CPU time already used, priority (if relevant), number of I/O requests performed.</strike>
- <strike>MEM - shows the current usage of memory space.</strike>
- <strike>LOAD - loads a program or job file into the simulator, and will also include the allocation of the program’s PCB and memory space.</strike>
- <strike>EXE - lets the simulation run on its own. The user can also specify the number of cycles to run before pausing. If there are no processes in the ready queue that are waiting to be scheduled, EXE will return to the command interface.</strike>
- <strike>RESET – allows the user to manually reset the simulator. All unfinished processes are terminated and the simulator clock returns to zero.</strike>
- <strike>EXIT – allows the user to end and exit the simulator.</strike>

<strike>2.3 Basic Operations
Four basic operations are required in this project program. They are calculation (processing), I/O simulation, yielding (interrupts), and output. Job files will contain the random sequence of these operations or program instructions to simulate activity of a program. Each job or simulated program in the text file will be like a “script”; text lines read one after another by the simulator. These operations or instructions will be stored into an array of strings along with any parameters that follow them. The number of cycles represents a “run-time” or the length of all the instructions and math operations to be simulated for that particular program.</strike>

<strike>2.4 User Interface
This program/simulator will incorporate a user interface so that he/she can control the flow of the operating system and observe the “running” of it for testing purposes. They will accomplish this by a load and an exe command. The load command will load a program or job file into the simulator and will also include the allocation of the program’s PCB and memory space. The exe command will let the simulator run on its own. The user can also specify the number of cycles to run before pausing.</strike>

<strike>2.5 Graphical User Interface
The GUI will display real-time statistics or data on all currently running processes. The PCB information of the jobs that are in memory and in the run state will be shown in some sort of GUI data table. This GUI table is always “refreshing” or updating as the simulator runs on its own or steps through the code.</strike>

2.6 Job and Program Files
<strike>“Job files” are text files that open and load the mock programs found in “program files”. The lines in a job file will be LOAD commands specifying the exact cycle time a process (in the program files) is to be loaded <strike>and put into the NEW queue or state</strike>. Also, the job file contains the string name of the process so it can be identified to the user. <strike>The EXE command is the last line in each job file script and will cause the simulator to load all other lines and information, storing it into the proper fields of the PCB.</strike>
<strike>Once loaded, the first line in a program file will input the process’ memory requirement (possibly an INT value). The following lines will be the “script” of operations: CALCULATE, I/O, YIELD, and OUT:</strike></strike>
- <strike>CALCULATE – When this is read in, the simulator will run the process in the run state for the number of cycles specified as a parameter</strike>
- <strike>I/O – The simulator reads in this command, but a random number generator creates the parameter value in the main code or simulator (likely an INT value). This will put the process in the blocked state.</strike>
- <strike>YIELD – This command yields from or pauses the running process, halts its accumulating cycle time and gives priority to another process (possibly through the use of round-robin scheduling). These can act as random interrupts in the system.</strike>
- <strike>OUT – This will print out a message to the screen, possibly to indicate which process is running and all of its PCB information. This is similar to the user entered PROC command, but is generated by the system.</strike>

<strike>2.7 Scheduling Algorithm
Project team must choose and implement a scheduling algorithm. Easiest solution would be to go with Round Robin approach due to the ease of implementation, however other solutions also will be most accepted. For advanced projects one may implement several solutions and compare their</strike>

<strike>2.8 Process States
The state implementation of our operating system simulator is as follows:
- NEW – The program or process is being created or loaded (but not yet in memory).
- READY – The program is loaded into memory and is waiting to run on the CPU.
- RUN – Instructions are being executed (or simulated).
- WAIT (BLOCKED) – The program is waiting for some event to occur (such as an I/O completion).
- EXIT – The program has finished execution on the CPU (all instructions and I/O complete) and leaves memory.</strike>

<strike>2.9 Memory Management
Memory Management will be implemented by keeping a running total of all processes in main memory by not letting the overall memory size exceed 256 kB. This value which is to be compared against the memory requirements of newly arrived processes. If total memory minus used memory (e.g. 256 kB – 180 kB = 76kB) is more than the newly arrived job’s memory requirement, it may enter the READY state (queue). Otherwise, the process would be entered into a waiting queue.</strike>

<strike>2.10 Event Processing
To simulate the features of the Operating System simulator an event driven paradigm will be employed. Any event waiting for service will be represented by an ECB (i.e. Event Control Block) and will wait for service on the EventQueue. Content of the ECB will differ according to the event type. EventQueue will be a Priority Queue implemented as a Binary Heap data structure.
Each ECB will be created by the appropriate handler such as the I/O Scheduler, <strike>CPU (executing interruptible instructions), Process Scheduling, and so on</strike>. In the case of I/O, the Interrupt Processor will monitor the EventQueue for events matching the current clock time and service those events accordingly. Long and Short Term Schedulers (e.g. Dispatcher) will monitor the EventQueue for events matching the current clock time and service those events accordingly.</strike>

3. Operating System Classes and Methods
Please find below minimum required functions for specific OS components.

Command Interface: Executes commands to the Operating System and the simulators.
proc()
<strike>mem()
load()
exe()
reset()</strike>
promptUser()

Scheduler: Schedules processes according to round robin algorithm with a time quantum of 10 cycles.
<strike>insertPCB()
removePCB()</strike>
getState()
setState()
<strike>getWait()</strike>
setWait()
<strike>getArrival()
setArrival()</strike>
getCPUTime()
setCPUTime()

<strike>
ExecutionQueue: Encapsulates the queue structure used by Scheduler.
enQueue()
deQueue()
</strike>

<strike>
CPU: Aggregates and supervises Clock, CacheMemory, InterruptProcessor, and eventQueue.
advanceClock()
detectInterrupt()
detectPreemption()
</strike>

<strike>
Clock: Simulates a CPU clock.
execute()
getClock()
</strike>

<strike>
InterruptProcessor: Simulates an interrupt processor.
signalInterrupt()
addEvent()
getEvent()
</strike>

<strike>EventQueue: Encapsulates the event queue used by interrupt system.
enQueue()
deQueue()</strike>

IOScheduler: Simulates I/O devices.
<strike>scheduleIO()</strike>
startIO()

<strike>IOBurst: Randomly determines IO burst times.
generateIOBurst ()</strike>

**Comment by StevenMHernandez** [comment]
@drbly @lobbypolice 
If either of you see things that you believe are done, please strike them out from the issue above (click edit near the top)

# 

```
To cross  out a line: <strike>this section will be crossed out</strike>
```

To cross  out a line: <strike>this section will be crossed out</strike>

**Comment by StevenMHernandez** [comment]
Also, if we are missing any important features, please create an issue here so that we don't forget to implement it.

**Comment by lobbypolice** [comment]
scheduling algorithm is done right?

**Comment by drbly** [comment]
i don't think so. every time i add something to it i take another look at the pdf and see that it is wrong. i am working on the ExecutionQueue, which is supposed to hold the queue for the scheduler. so i guess the scheduler just gets processes from the queue and moves them to the CPU?

**Comment by lobbypolice** [comment]
i dunno i thought the max cpu time was the round robin type algorithm and it will automatically move them out for the next process? or does it not do that right now
